### DISCUSSION TITLE
Cross-regex ambiguities

### DISCUSSION THREAD

**Discussion by mehmetoguzderin** [tree-sitter/tree-sitter#1430] [Q&A] [answered]
(Apologies if this was a topic of past discussion.)

Tree-sitter does not warn if regex patterns cause ambiguity in the grammar. A comprehensive solution might be hard to achieve, but this specific case we ran into is just a top-level check.

Or is there already a way to enable such a check within Tree-sitter?

Thanks!

See: https://github.com/gpuweb/gpuweb/issues/2169

**Comment by ahelwer** [comment]
Tree-sitter's rules of lexical precedence are well-defined, can you provide an example of such an ambiguity? Edit: I see the link, reading it now. Here are the lexical precedence rules: https://tree-sitter.github.io/tree-sitter/creating-parsers#conflicting-tokens

You'd want a warning if it falls back to the ordering rule, or what?

**Reply by mehmetoguzderin** [reply]
Right, maybe a way to mark as not intentional like this comment: https://github.com/tree-sitter/tree-sitter/discussions/1430#discussioncomment-1443862

**Comment by maxbrunsfeld** [comment, marked as answer]
Yeah, Tree-sitter is aware of which regexes "conflict" with one another, but it doesn't issue any kind of warning, because this occurs intentionally in many grammars.

Maybe there is some particular case where we can determine that it's not intentional, and issue a conflict error? I'm not sure what the rule would be though.

**Reply by mehmetoguzderin** [reply]
I am not very familiar, so it would be wrong to suggest a shape for the change, but having the option to surface it would be great!
